### PR TITLE
[wip] local development environment for pilot

### PR DESCRIPTION
**Pull request by kyessenov**
This let you spin up local proxy, agent, discovery, and mixer, and issue requests as if through the client-side proxy.

E.g.:
```
go run test/local/main.go -logtostderr -envoy bazel-genfiles/proxy/envoy/envoy -v 2
curl localhost:8000  --header 'Host: dev.local'
```

This should also work on mac (without mixer) if you use darwin build of Envoy.

### CODE REVIEW

**Comment by kyessenov** [comment]
Confirmed that it also works on mac with upstream envoy.

File: bin/push-docker [hunk 1]
```diff
@@ -51,7 +51,7 @@ done
 bin=$(bazel info bazel-bin)
 bazel build --output_groups=static //cmd/... //test/...
 \cp -f  "${bin}/test/client/client.static" docker/client
-\cp -f  "${bin}/test/server/server.static" docker/server
+\cp -f  "${bin}/test/server/main/main.static" docker/server
```
Comment (rshriram) [inline bin/push-docker]: has this changed in our repos? or is this your local copy?

**Comment by kyessenov** [comment]
Yeah, it's using non-transparent mode in this setup. It gets more complicated with server-side which is why this is WIP.

File: bin/push-docker (hunk 1 above)
Comment (kyessenov) [inline bin/push-docker]: This change is just in this PR.

File: proxy/envoy/discovery.go [hunk 2]
```diff
@@ -194,13 +195,15 @@ type DiscoveryServiceOptions struct {
...
 	EnableCaching   bool
+	EnableLocal     bool
 }
 
 // NewDiscoveryService creates an Envoy discovery service on a given port
 func NewDiscoveryService(ctl model.Controller, configCache model.ConfigStoreCache,
 	environment proxy.Environment, o DiscoveryServiceOptions) (*DiscoveryService, error) {
 	out := &DiscoveryService{
 		Environment: environment,
+		useVirtual:  !o.EnableLocal,
```
Comment (andraxylia) [inline proxy/envoy/discovery.go]: I would rename the useVirtual flag to enableLocal and use the same naming everywhere.

If I understand correctly, this distinguishes the case where iptable redirection is used, so I am not sure if the word "virtual" is the most appropriate.

File: proxy/envoy/discovery.go [hunk 3]
```diff
@@ -219,13 +222,16 @@ func NewDiscoveryService(ctl model.Controller, configCache model.ConfigStoreCach
...
+
+	if ctl != nil {
+		serviceHandler := func(s *model.Service, e model.Event) { out.clearCache() }
+		if err := ctl.AppendServiceHandler(serviceHandler); err != nil {
+			return nil, err
+		}
+		instanceHandler := func(s *model.ServiceInstance, e model.Event) { out.clearCache() }
+		if err := ctl.AppendInstanceHandler(instanceHandler); err != nil {
+			return nil, err
+		}
```
Comment (andraxylia) [inline proxy/envoy/discovery.go]: Is ctl == nil an error case? If yes, check it in the beginning and return an error.

If not, is it worth logging it.

File: test/local/main.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	discoveryPort int
+	configpath    string
+	binarypath    string
+	service       *model.Service
+)
+
+func init() {
+	flag.IntVar(&baseHTTPPort, http, 9000, "Base HTTP/1.1 port")
+	flag.IntVar(&baseGRPCPort, grpc, 9100, "Base gRPC port")
+	flag.IntVar(&n, "n", 10, "Number of instances")
```
Comment (andraxylia) [inline test/local/main.go]: Qualify what instances is this about. For example:
"Number of proxy instances" or "Envoy instance" or "app instances"

File: test/local/main.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+func (_ registry) Instances(hostname string, ports []string, tagsList model.TagsList) []*model.ServiceInstance {
+	if hostname != host {
+		return nil
+	}
+
+	out := make([]*model.ServiceInstance, 0)
+
+	for _, port := range ports {
+		var base int
+		if port == http {
```
Comment (andraxylia) [inline test/local/main.go]: Use a switch instead of if/else.

**Comment by linsun** [comment]
Very nice.  Does it make sense to have the instruction added to here?  https://github.com/istio/pilot/blob/master/doc/testing.md

**Comment by kyessenov** [comment]
It's not ready and not high priority for me. I'm putting it out since it's useful for debugging.

**Comment by rshriram** [comment]
Closing temporarily to reduce load on bot.

**Pull request closed without merging**